keps/sig-etcd/4647-cluster-feature-gate/README.md

```diff
@@ -0,0 +1,548 @@
+}
+```
+
+#### New Backend Schema Changes
+We will save the values of cluster feature_gates in the `Cluster` bucket in the backend.
+
+#### Start a New Cluster
+
+Fig. a above shows how the consensus of feature gates is reached for a new cluster:
+1. When the cluster starts, before `ClusterVersion` is set, the cluster feature gate would be `nil`, and all features would be considered disabled. 
+1. Before the leader send an update message of the `ClusterVersion`, the leader will send a `ClusterFeatureGateSetRequest` message with empty `feature_gates` to reset any existing cluster feature gate. This is to make sure the feature gate is always compatible with the `ClusterVersion`.
+1. Whenever the `ClusterVersionSet` is applied in a server, the member will propose the cluster feature gate values based on its own flags and the `ClusterVersion`.
+1. Whenever the leader receives the `member.Attributes.proposed_feature_gates` from a member, it would try to do `UpdateClusterFeatureGate` by reconciling the `proposed_feature_gates` from all the members: 
+   * if the `proposed_feature_gates` from any member is missing, the cluster feature gate will be empty, i.e. all features will be disabled.
+     * [NOTE] the leader would wait for all the members to send in their `proposed_feature_gates` instead of just initialize the feature gate to the default values at the `ClusterVersion`, because different members might have different patch versions, and may not have the same default values. We do not want the feature values to flip-flop when the leader changes.
+     * [NOTE] the leader would wait for all the members to send in their `proposed_feature_gates` instead of setting the feature gates based on the `proposed_feature_gates` it has received so far, because it generally better to move the feature gate from empty to the final step in one step, instead of changing it in N(member) number of steps and beining also dependent on the order of the messages the leader receives.
+   * a feature is disabled if the feature is proposed to be `false` by any member
+   * a feature is enabled only if the feature is proposed to be `true` by all members
+   * a feature is disabled if it is missing from the proposal of any member.
+1. The leader will send `ClusterFeatureGateSetRequest` through raft to set the final values of the cluster feature gate to all members. 
+
+One drawback of this approach is that there is a period before the leader receives the proposals from all members when no feature would be available. The duration of this period is expected to be `O(time to set ClusterVersion)`. We think that is acceptable because if the time to set `ClusterVersion` is exceptionally long, the `ClusterVersion` would stay at the fallback version `3.0`, for which a lot of etcd capabilities would not be available anyway. It is better to wait than risking splitting minds about what the default values should be when relying on falling back to default values of the `ClusterVersion`.
```

Comment:
We should not consider cluster working until the cluster version is established. We could consider defining a cluster boostrap procedure that clarifies that cluster should not serve client traffic until it's ready.

Reply:
cc @ahrtr

Reply:
Currently the cluster version is established only after the leader can get the version from all members. considering cluster working until the cluster version is established would break HA assumption, meaning the cluster would not be considered working until all members are bootstrapped, instead of the majority. This would be a big change.